src/shop/api.py

```diff
@@ -331,3 +348,70 @@ def update_out_of_stock_forecast(product_id):
     product_obj = models.Product.objects.get(id=product_id)
     product_obj.out_of_stock_forecast = predict_quantity(product_id, target=0)
     product_obj.save()
+
+
+def order_from_supplier(product_id, qty, supplier_id=None):
+    """Orders the cheapest product from a supplier."""
+    products = models.SupplierProduct.objects.filter(product_id=product_id)
+
+    if supplier_id is not None:
+        products = products.filter(supplier_id=supplier_id)
+
+    # A product can be associated with several different supplier products from
+    # the same supplier. Supplier products are also most often sold in batches,
+    # which means you usually cannot purchase the exact amount you need.
+    # Following section of the code takes care of calculating the minimum cost
+    # of each product while taking `qty` into account. It will then try to
+    # purchase the cheapest one and if not possible (for example out of stock),
+    # continue onto next one.
+    def minimum_qty(sp):
+        return math.ceil(qty / sp.qty)
+
+    def cost(sp):
+        # Calculate the minimum quantity of the supplier product that needs to
+        # be purchased in order to reach `qty`.
+        return minimum_qty(sp) * sp.qty * sp.unit_price
+
+    products = sorted(products, key=cost)
+    for product in products:
+        supplier = product.supplier
+        supplier_api = suppliers.get_supplier_api(supplier.internal_name)
+        try:
+            supplier_api.order_product(product.sku, minimum_qty(product))
+            return product
+        except SupplierAPIException:
+            # Log the error and try the next product.
+            log.warning('Failed to order product SKU %d from %s.',
+                        product.sku, supplier.internal_name)
+    else:
+        msg = 'Could not order {}.'.format(product.sku)
+        raise exceptions.APIException(msg)
```

Comment:
`else` won't be executed if you do a `break` in your loop. https://docs.python.org/3/tutorial/controlflow.html#break-and-continue-statements-and-else-clauses-on-loops

I do not know however how I feel about translating internal exceptions. I know that I'm displaying their messages in some places (exceptions from the stock taking module) in the admin panel, but it feels kind of wrong. I'll think about it and get back to you.